src/fs.rs

```diff
@@ -143,22 +143,45 @@ impl BiosParameterBlock {
         bpb.total_sectors_32 = rdr.read_u32::<LittleEndian>()?;
 
         // sanity checks
-        if bpb.bytes_per_sector < 512 {
+        if bpb.bytes_per_sector.count_ones() != 1 {
             return Err(Error::new(ErrorKind::Other, "invalid bytes_per_sector value in BPB"));
-        }
-        if bpb.sectors_per_cluster < 1 {
+        } else if bpb.bytes_per_sector < 512 {
+            return Err(Error::new(ErrorKind::Other, "invalid bytes_per_sector value in BPB"));
+        } else if bpb.bytes_per_sector > 4096 {
+            return Err(Error::new(ErrorKind::Other, "invalid bytes_per_sector value in BPB"));
+        } 
+
+        if bpb.sectors_per_cluster.count_ones() != 1 {
+            return Err(Error::new(ErrorKind::Other, "invalid sectors_per_cluster value in BPB"));
+        } else if bpb.sectors_per_cluster < 1 {
+            return Err(Error::new(ErrorKind::Other, "invalid sectors_per_cluster value in BPB"));
+        } else if bpb.sectors_per_cluster > 128 {
             return Err(Error::new(ErrorKind::Other, "invalid sectors_per_cluster value in BPB"));
         }
+
+        // bytes per sector is u16, sectors per cluster is u8, so guaranteed no overflow in multiplication
+        let bytes_per_cluster : u32 = bpb.bytes_per_sector as u32 * bpb.sectors_per_cluster as u32; 
+        let maximum_bytes_per_cluster : u32 = u32::max_value();
```

Comment:
OK, updated to remove hard maximum, and to warn!() when >32k.
A warning is appropriate because a cluster size > 32k will be incompatible with some implementations of FAT.